tests/gui/test_widgets.py

```diff
@@ -5812,3 +5938,90 @@ def test_SenderIcon_set_normal_styles_purple_for_authenticated_user(mocker):
+    session.add(journalist)
+    session.commit()
+
+    # Add the record that says the journalist saw the message
+    session.add(db.SeenMessage(message_id=message.id, journalist_id=journalist.id))
+    session.commit()
+
+    # Create the MessageWidget for the message above
+    controller = mocker.MagicMock(authenticated_user=journalist)
+    cv = ConversationView(source, controller)
+    cv.update_conversation(source.collection)
+    message_widget = cv.current_messages[message.uuid]
+
+    # Get the tool tip text and compare it to the expected result
+    assert message_widget.check_mark.toolTip() == "dellsberg"
+
+    # Update the conversation again after another journalist has seen the same message
+    second_journalist = factory.User(username="journalist")
+    session.add(second_journalist)
+    session.commit()
+
+    # Add the record that says the second journalist saw the message
+    session.add(db.SeenMessage(message_id=message.id, journalist_id=second_journalist.id))
+    session.commit()
+
+    # Update the source collection
+    cv.update_conversation(source.collection)
+
+    m = mocker.MagicMock()
+    dict = {journalist.username: journalist, second_journalist.username: second_journalist}
+    m.__getitem__.side_effect = dict.__getitem__
+    assert message_widget.check_mark.toolTip() == "journalist,\ndellsberg"
+
+
+def test_ConversationView_updates_reply_seenby_tooltip(mocker, session):
+    """
+    Ensure the tooltip displays the usernames of the users who have seen the replies
+    in the correct order.
+    """
+    # Create a source, message, and a journalist that will see that message
+    source = factory.Source()
+    session.add(source)
+    reply = factory.Reply(source=source)
+    session.add(reply)
+    journalist = factory.User(username="dellsberg")
+    session.add(journalist)
+    session.commit()
+
+    # Add the record that says the journalist saw the message
+    session.add(db.SeenReply(reply_id=reply.id, journalist_id=journalist.id))
+    session.commit()
+
+    # Create the MessageWidget for the message above
+    controller = mocker.MagicMock(authenticated_user=journalist)
+    cv = ConversationView(source, controller)
+    cv.update_conversation(source.collection)
+    reply_widget = cv.current_messages[reply.uuid]
+
+    # Get the tool tip text and compare it to the expected result
+    assert reply_widget.check_mark.toolTip() == "dellsberg"
+
+    # Update the conversation again after another journalist has seen the same message
+    second_journalist = factory.User(username="journalist")
+    session.add(second_journalist)
+    session.commit()
+
+    # Add the record that says the second journalist saw the message
+    session.add(db.SeenReply(reply_id=reply.id, journalist_id=second_journalist.id))
+    session.commit()
+
+    # Update the source collection
+    cv.update_conversation(source.collection)
+
+    assert reply_widget.check_mark.toolTip() == "journalist,\ndellsberg"
```

Comment:
Stylistic comment: I'm not sure I find the comma useful if we're presenting the tool tip as a vertical list of names. YMMV